core/sbt-cloudflow/src/main/scala/cloudflow/sbt/CloudflowFlinkPlugin.scala

```diff
@@ -0,0 +1,240 @@
+      buildStreamletDescriptors(detectedStreamlets, dockerImageName)
+    }.value,
+
+    buildOptions in docker := BuildOptions(
+      cache = true,
+      removeIntermediateContainers = BuildOptions.Remove.OnSuccess,
+      pullBaseImage = BuildOptions.Pull.IfMissing
+    ),
+
+    cloudflowStageAppJars := Def.taskDyn {
+      Def.task {
+        val stagingDir = stage.value
+        val projectJars = (Runtime / internalDependencyAsJars).value.map(_.data)
+        val depJars = (Runtime / externalDependencyClasspath).value.map(_.data)
+
+        val appJarDir = new File(stagingDir, AppJarsDir)
+        val depJarDir = new File(stagingDir, DepJarsDir)
+        projectJars.foreach { jar ⇒
+          IO.copyFile(jar, new File(appJarDir, jar.getName))
+        }
+        depJars.foreach { jar ⇒
+          if (jar.name startsWith "cloudflow-runner-") {
+            IO.copyFile(jar, new File(depJarDir, "cloudflow-runner.jar"))
+          } else IO.copyFile(jar, new File(depJarDir, jar.getName))
+        }
+      }
+    }.value,
+
+    imageNames in docker := {
+      // NOTE: only use the default repository name ("lightbend") if a registry
+      //       has been set but no repository has been set.
+      //
+      //       When using the Openshift docker registry, the repository MUST
+      //       already exist as a K8s namespace for the fallback to work.
+      //       We fall back to "lightbend" because we know that Cloudflow will
+      //       be installed in that namespace
+      val registry = cloudflowDockerRegistry.value
+      val namespace = cloudflowDockerRepository.value.orElse(registry.map(_ ⇒ "lightbend"))
+
+      cloudflowFlinkDockerImageName.value
+        .map { imageName ⇒
+          ImageName(
+            registry = registry,
+            namespace = namespace,
+            repository = imageName.name,
+            tag = Some(imageName.tag)
+          )
+        }
+        .toSeq
+    },
+
+    dockerfile in docker := {
+      // NOTE !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+      // The UID and GID of the `jboss` user is used in different parts of Cloudflow
+      // If you change this, you have to make sure that all references to this value are changed
+      // - fsGroups on streamlet pods uses the GID to make volumes readable
+      val userInImage = "185" // default non-root user in the spark image
+      val userAsOwner: String ⇒ String = usr ⇒ s"$usr:cloudflow"
+
+      // this triggers side-effects, e.g. files being created in the staging area
+      cloudflowStageAppJars.value
+
+      val appDir: File = stage.value
+      val appJarsDir: File = new File(appDir, AppJarsDir)
+      val depJarsDir: File = new File(appDir, DepJarsDir)
+
+      // TODO: make it so we only build docker images when the blueprint is valid
+      val streamletDescriptorsLabelName = "com.lightbend.cloudflow.streamlet-descriptors"
+      // pack all streamlet-descriptors into a Json array
```

Comment:
It might be smart to put this inside a root object that has a property that contains the array instead, for extensibility.

Reply:
done ..